text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,629 @@
+2. Annotate the HIR map accessor methods so that they add appropriate
+   read/write edges.
+
+This will integrate with the "default visitor" described under
+procedure nodes. This visitor can hand off just an opaque id for each
+item, requiring the pass itself to go through the map to fetch the
+actual HIR, thus triggering a read edge (we might also bake this
+behavior into the visitor for convenience).
+
+### Persisting the graph
+
+Once we've built the graph, we have to persist it, along with some
+associated information. The idea is that the compiler, when invoked,
+will be supplied with a directory. It will store temporary files in
+there. We could also consider extending the design to support use by
+multiple simultaneous compiler invocations, which could mean
+incremental compilation results even across branches, much like ccache
+(but this may require tweaks to the GC strategy).
+
+Once we get to the point of persisting the graph, we don't need the
+full details of the graph. The process nodes, in particular, can be
+removed. They exist only to create links between the other nodes. To
+remove them, we first compute the transitive reachability relationship
+and then drop the process nodes out of the graph, leaving only the HIR
+nodes (inputs) and IR nodes (output).  (In fact, we only care about
+the IR nodes that we intend to persist, which may be only a subset of
+the IR nodes, so we can drop those that we do not plan to persist.)
+
+For each HIR node, we will hash the HIR and store that alongside the
+node. This indicates precisely the state of the node at the time.
+Note that we only need to hash the HIR itself; contextual information
+(like `use` statements) that are needed to interpret the text will be
+part of a separate HIR node, and there should be edges from that node
+to the relevant compiler data structures (such as the name resolution
+tables).
+
+For each IR node, we will serialize the relevant information from the
+table and store it. The following data will need to be serialized:
+
+- Types, regions, and predicates
+- ADT definitions
+- MIR definitions
+- Identifiers
+- Spans
+
+This list was gathered primarily by spelunking through the compiler.
+It is probably somewhat incomplete. The appendix below lists an
+exhaustive exploration.
+
+### Reusing and garbage collecting artifacts
+
+The general procedure when the compiler starts up in incremental mode
+will be to parse and macro expand the input, create the corresponding
+set of HIR nodes, and compute their hashes. We can then load the
+previous dependency graph and reconcile it against the current state:
+
+- If the dep graph contains a HIR node that is no longer present in the
+  source, that node is queued for deletion.
+- If the same HIR node exists in both the dep graph and the input, but
+  the hash has changed, that node is queued for deletion.
+- If there is a HIR node that exists only in the input, it is added
+  to the dep graph with no dependencies.
```

Comment:
This sounds like we should try to keep and update the old dependency graph. I don't think that's a good idea, since the old graph has undergone some irreversible transformations during persisting and I don't see any benefit in it. Just use the old graph for deleting invalidated cache entries and build a new dependency graph from scratch.